python/models/hierarchical_filtering_proxy_model.py

```diff
@@ -329,104 +333,122 @@ def setSourceModel(self, model):
         prev_source_model = self.sourceModel()
         if prev_source_model:
             prev_source_model.rowsInserted.disconnect(self._on_source_model_rows_inserted)
+            prev_source_model.rowsRemoved.disconnect(self._on_source_model_rows_removed)
             prev_source_model.dataChanged.disconnect(self._on_source_model_data_changed)
             prev_source_model.modelAboutToBeReset.disconnect(self._on_source_model_about_to_be_reset)
 
-        # clear out the various caches:
-        self._dirty_all_accepted()
+        # clear out the rejection cache:
+        self._clear_rejection_cache()
 
         # call base implementation:
         QtGui.QSortFilterProxyModel.setSourceModel(self, model)
 
         # connect to the new model:
         if model:
             model.rowsInserted.connect(self._on_source_model_rows_inserted)
+            model.rowsRemoved.connect(self._on_source_model_rows_removed)
             model.dataChanged.connect(self._on_source_model_data_changed)
             model.modelAboutToBeReset.connect(self._on_source_model_about_to_be_reset)
 
     # -------------------------------------------------------------------------------
     # Private methods
 
-    def _is_child_accepted_r(self, idx, parent_accepted):
+    def _update_rejection_cache(self, from_index):
         """
-        Recursively check children to see if any of them have been accepted.
+        Update the rejection cache from the given index.
 
-        :param idx:             The model index whose children should be checked
-        :param parent_accepted: True if a parent item has been accepted
-        :returns:               True if a child of the item is accepted by the filter
-        """
-        model = idx.model()
+        The cache at the given index has to be cleared before calling this method
+        to force a refresh.
 
-        # check to see if any children of this item are known to have been accepted:
-        cached_value = self._child_accepted_cache.get(idx)
+        :param from_index: A :class:`QModelIndex` instance.
+        """
+        if not from_index or not from_index.isValid():
+            return
+        # Do we have a value from a previous iteration?
+        cached_value = self._rejected_cache.get(from_index)
```

Comment:
With the suggestion above:
```
if self._rejected_cache.is_cached(from_index):
   return self._rejected_cache.is_accepted(from_index)
```

Reply:
I personally find  `self._rejected_cache.is_accepted` confusing. I agree that `return not cached_value` is not great either... There is a notion of being _implicitly_ accepted for some items which are not in the cache. Let me see if I can make that clearer.